app/src/main/java/com/github/gotify/PicassoHandler.java

```diff
@@ -0,0 +1,118 @@
+package com.github.gotify;
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import com.github.gotify.api.CertUtils;
+import com.github.gotify.api.ClientFactory;
+import com.github.gotify.client.ApiClient;
+import com.github.gotify.client.api.ApplicationApi;
+import com.github.gotify.client.model.Application;
+import com.github.gotify.messages.provider.MessageImageCombiner;
+import com.squareup.picasso.OkHttp3Downloader;
+import com.squareup.picasso.Picasso;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import okhttp3.Cache;
+import okhttp3.OkHttpClient;
+
+/**
+ * Copyright (C) 2020 Felix Nüsse Created on 01.06.20 - 20:03
+ *
+ * <p>Edited by: Felix Nüsse felix.nuesse(at)t-online.de
+ *
+ * <p>gotify-android
+ *
+ * <p>This program is released under the MIT License
+ */
+public class PicassoHandler {
+
+    public static int PICASSO_CACHE_SIZE = 50 * 1024 * 1024; // 50 MB
+    public static String PICASSO_CACHE_SUBFOLDER = "picasso-cache";
+
+    private Context context;
+    private Settings settings;
+
+    private Cache picassoCache;
+
+    private Picasso picasso;
+    private Map<Integer, String> appIdToAppImage = new ConcurrentHashMap<>();
+
+    public PicassoHandler(Context context, Settings settings) {
+        this.context = context;
+        this.settings = settings;
+
+        picasso = makePicasso();
+    }
+
+    private void prepareCache() {
+        if (picassoCache == null) {
```

Comment:
I think this is not needed, makePicasso should only be called once. The cache should be instantiated inside the constructor.